### PR TITLE
Muon HLT IterL3 code

### PR DESCRIPTION
**Pull request by folguera**
This PR includes the following changes: 
 - Re-design of the HLTMuonL2SelectorForIO to use track links and to apply a L3 filter on the fly.
 - Some modifications on TSGForOI to simplify the code and to decide whether to run with hits or not depending on the L2 parameters.
- The updated and optimised customisation function. 
- Some naming conventions and cosmetics in the previous mentioned classes.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

@folguera please a link to some slides describing the changes (I expect that it was presented in TSG already)

**Comment by folguera** [comment]
The slides were indeed presented at the TSG: https://indico.cern.ch/event/605949/contributions/2442701/subcontributions/218440/attachments/1401603/2139430/Muon_TSG_25_01_2017.pdf

I'm fixing the errors, locally I've not seen those and commit again.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by davidlange6** [comment]
assign hlt

**Comment by silviodonato** [comment]
Apart from my minor comments, the code is ok and working: I checked that with the current HLT configuration the number of accepted event for each HLT path does not change.

**Comment by folguera** [comment]
Hi, @silviodonato I do not see any comments on the code, am I looking at the wrong place?

**Comment by silviodonato** [comment]
Right, I didn't post them!

**Comment by folguera** [comment]
I'm missing the second comment, don't launch the test yet, please

**Comment by slava77** [comment]
@folguera 
do you plan to make more commits here still?

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc [hunk 1]
```diff
@@ -1,78 +1,108 @@
...
+  for (unsigned int il2=0; il2 != l2muonH->size(); ++il2){
+    reco::TrackRef l2muRef(l2muonH, il2);
+    bool re_do_this_L2=true;
+    
+    for (unsigned int il3=0; il3 != l3muonH->size(); ++il3){
+      reco::RecoChargedCandidateRef cand(l3muonH,il3);
+      reco::TrackRef tk = cand->track();
+      
+      bool useThisLink = false;
+      for (unsigned int l(0); l <links->size(); ++l){
```
Comment (silviodonato) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: I think it's better to use " l <links->size() && !useThisLink" and remove the subsequent line
"if (useThisLink) continue;"

**Comment by folguera** [comment]
@slava77 no more commits from my side (unless further comments/changes are suggested)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by silviodonato** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17325	3334114
- all changes are HLT-specific, reco is not affected. The code changes look OK after a somewhat cursory inspection.
- jenkins tests pass and comparisons with baseline show no differences

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc [hunk 2]
```diff
@@ -1,78 +1,106 @@
...
+      
+      bool useThisLink = false;
+      for (unsigned int l(0); l<links->size() && !useThisLink; ++l){
+	const reco::MuonTrackLinks* link = &links->at(l);
+	
+	// Check if the L3 link matches the L3 candidate
+	const reco::Track& globalTrack = *link->globalTrack();
+	float dR2 = deltaR2(tk->eta(),tk->phi(),globalTrack.eta(),globalTrack.phi());
+	float dPt = std::abs(tk->pt() - globalTrack.pt())/tk->pt();
+	if (dR2 < 0.02*0.02 and dPt < 0.001) {
```
Comment (davidlange6) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: @folguera - better to write this as 
if ( dR2< 0.02*0.02 and std::abs(tk->pt() - globalTrack.pt()) < 0.001*tk->pt() ) {....

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc [hunk 3]
```diff
@@ -1,78 +1,106 @@
...
+	  useThisLink=true;
```
Comment (davidlange6) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: you can also stop the loop once this is true?

Comment (davidlange6) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: whoops, I'm lost with the } - never mind that looping comment

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc [hunk 4]
```diff
@@ -1,78 +1,106 @@
...
+	if (l2muRef==staTrack) re_do_this_L2=false; 
+      
+	// Check the quality of the reconstructed L3, if poor quality, we will re-use this L2
+	if (staTrack==l2muRef && applyL3Filters_) {
+	  re_do_this_L2 = true;
+	  const reco::Track& globalTrack = *link->globalTrack();
+	  if (globalTrack.numberOfValidHits()< min_Nhits_) continue;  	// cut on number of hits
+	  if (globalTrack.normalizedChi2() > max_NormalizedChi2_ ) continue;		//normalizedChi2 cut
+	  if (globalTrack.hitPattern().numberOfValidMuonHits() < min_NmuonHits_ ) continue;  	//min muon hits cut
+	  if (std::abs(globalTrack.pt()-l2muRef->pt())/globalTrack.pt() > max_PtDifference_) continue; // pt difference
```
Comment (davidlange6) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: similar comment to above - the /pt() can be avoided (faster and safer)

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc [hunk 5]
```diff
@@ -48,235 +52,249 @@ TSGForOI::~TSGForOI(){
...
-	return theSF;
+double TSGForOI::calculateSFFromL2(const reco::TrackRef track){
+
+  double theSF=1.0;
+  //	L2 direction vs pT blowup - as was previously done:
+  //	Split into 4 pT ranges: <pT1_, pT1_<pT2_, pT2_<pT3_, <pT4_: 13,30,70
+  //	Split into 2 eta ranges for the middle two pT ranges: 1.0,1.4
+  if (track->pt()<=pT1_) theSF=SF1_;
+  if (track->pt()>pT1_ && track->pt()<=pT2_){
+    if (std::abs(track->eta())<=eta1_) theSF=SF3_;
```
Comment (davidlange6) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: please cache the result of abs(track->eta() ) here to use in the next lines

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc (hunk 2 above)
Comment (folguera) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: fixed

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc (hunk 3 above)
Comment (folguera) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: this was the comment also from Silvio, it is fixed.

File: RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc (hunk 4 above)
Comment (folguera) [inline RecoMuon/L3TrackFinder/src/HLTMuonL2SelectorForL3IO.cc]: fixed

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc (hunk 5 above)
Comment (folguera) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc]: done

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17325	2ed64aa
- after some technical code improvements since the last signoff https://github.com/cms-sw/cmssw/compare/3334114e72f13bba665058d2522b6169b2dba62a...2ed64aa88730b5b1acf94a758394fa75543761d7
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by silviodonato** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**